crates/codegen/parser/src/code_generator.rs

```diff
@@ -0,0 +1,318 @@
+    }
+
+    pub fn scanner_functions(&self) -> Vec<String> {
+        self.scanners
+            .iter()
+            .map(|(name, scanner)| {
+                let function_name = format_ident!("scan_{}", name.to_snake_case());
+                let body = scanner.to_function_body();
+                format!(
+                    "{}\n{}",
+                    scanner.to_comment_string(),
+                    quote! {
+                        #[allow(unused_assignments, unused_parens)]
+                        pub(crate) fn #function_name(&self, stream: &mut Stream) -> bool {
+                            #body
+                        }
+                    }
+                )
+            })
+            .collect()
+    }
+
+    pub fn parser_functions(&self) -> Vec<String> {
+        self.parsers
+            .iter()
+            .map(|(name, parser)| {
+                let kind = format_ident!("{}", name);
+                let function_name = format_ident!("parse_{}", name.to_snake_case());
+                let body = parser.to_function_body();
+                format!(
+                    "{}\n{}",
+                    parser.to_comment_string(),
+                    quote! {
+                        #[allow(unused_assignments, unused_parens)]
+                        pub(crate) fn #function_name(&self, stream: &mut Stream) -> ParseResult {
+                            match #body {
+                        Pass{ node, error } => Pass{ node: cst::Node::top_level_rule(RuleKind::#kind, node), error },
+                                fail => fail
+                            }
+                        }
+                    }
+                )
+            })
+            .collect()
+    }
+
+    pub fn scanner_invocations(&self) -> Vec<String> {
+        self.scanners
+            .keys()
+            .map(|name| {
+                let production_kind = format_ident!("{}", name);
+                let function_name = format_ident!("scan_{}", name.to_snake_case());
+                let error_message = name;
+                quote! {
+                    ProductionKind::#production_kind => call_scanner(self, input, Language::#function_name, TokenKind::#production_kind, #error_message)
+                }
+                .to_string()
+            })
+            .collect()
+    }
+
+    pub fn parser_invocations(&self) -> Vec<String> {
+        self.parsers
+            .keys()
+            .map(|name| {
+                let production_kind = format_ident!("{}", name);
+                let function_name = format_ident!("parse_{}", name.to_snake_case());
+                quote! {
+                    ProductionKind::#production_kind => call_parser(self, input, Language::#function_name)
+                }
+                .to_string()
+            })
+            .collect()
+    }
+
+    pub fn token_kinds(&self) -> Vec<String> {
+        self.token_kinds.keys().cloned().collect()
+    }
+
+    pub fn rule_kinds(&self) -> Vec<String> {
```

Comment:
Nit: in all three `token_kinds()`, `rule_kinds()`, and `production_kinds()`, I suggest sorting the list before returning. Not only this makes the output more readable, it also makes the diffs consistent during code review.

Reply:
BTreeMap `keys()` are sorted IIUC.

Reply:
Aha! good catch. In that case, my comment only concerns `production_kinds()` as it is concatenating two lists. It produces unsorted output in the generated `kinds.rs` in this PR.